tb_db/crud.py

```diff
@@ -341,3 +343,128 @@ def create_miru_profiles(db: Session, miru_profiles_by_sample_id: dict[str, obje
+
+    existing_cgmlst_clusters = db.query(CgmlstCluster).all()
+    existing_cgmlst_cluster_ids = set([cluster.cluster_id for cluster in existing_cgmlst_clusters])
+
+    #cluster_id = cgmlst_cluster['cluster']
+    db_samples = []
+    added_cgmlst_cluster_ids = set()
+    for row in cgmlst_cluster:
+        sample_id = row['sample_id']
+        cluster_id = row['cluster']
+        if (cluster_id not in existing_cgmlst_cluster_ids) and (cluster_id not in added_cgmlst_cluster_ids):
+            db_cgmlst_cluster = CgmlstCluster(
+                cluster_id = cluster_id,
+            )
+            db.add(db_cgmlst_cluster)
+            db.commit()
+            added_cgmlst_cluster_ids.add(cluster_id)
+        select_cgmlst_cluster_stmt = select(CgmlstCluster).where(and_(CgmlstCluster.cluster_id == cluster_id, CgmlstCluster.valid_until == None))
+        db_cgmlst_cluster = db.scalars(select_cgmlst_cluster_stmt).one()
+        if sample_id not in existing_sample_ids:
+            logging.warning('cannot add cgmlst cluster for a sample that does not exist...')         
+        else:
+            select_sample_stmt = select(Sample).where(and_(Sample.sample_id == sample_id, Sample.valid_until == None))
+            sample = db.scalars(select_sample_stmt).one()
+            sample.cgmlst_cluster_id = db_cgmlst_cluster.id
+            db_samples.append(sample)
+            db.commit()
+            
+
+    #select_sample_stmt = select(Sample).where(and_(Sample.sample_id == sample_id, Sample.valid_until == None))
+    #sample = db.scalars(select_sample_stmt).one()
+    return db_samples
+
+### cgmlst
+def add_sample_to_cgmlst_cluster(db: Session, sample_id: str, cgmlst_cluster: dict[str, object]):
+
+    existing_samples = db.query(Sample).all()
+    existing_sample_ids = set([sample.sample_id for sample in existing_samples])
+
+    existing_cgmlst_clusters = db.query(CgmlstCluster).all()
+    existing_cgmlst_cluster_ids = set([cluster.cluster_id for cluster in existing_cgmlst_clusters])
+
+    cluster_id = cgmlst_cluster['cluster']
+    if (cluster_id not in existing_cgmlst_cluster_ids):
+        db_cgmlst_cluster = CgmlstCluster(
+            cluster_id = cluster_id,
+        )
+        db.add(db_cgmlst_cluster)
+        db.commit()
+
+    select_cgmlst_cluster_stmt = select(CgmlstCluster).where(and_(CgmlstCluster.cluster_id == cluster_id, CgmlstCluster.valid_until == None))
+    db_cgmlst_cluster = db.scalars(select_cgmlst_cluster_stmt).one()
+
+    if sample_id not in existing_sample_ids:
+        logging.warning('cannot add cgmlst cluster for a sample that does not exist...')
+        return None
+
+    else:
+        select_sample_stmt = select(Sample).where(and_(Sample.sample_id == sample_id, Sample.valid_until == None))
+        sample = db.scalars(select_sample_stmt).one()
+        sample.cgmlst_cluster_id = db_cgmlst_cluster.id
+        db.commit()
+        return sample
```

Comment:
We may need to call db.refresh() to update the `sample.id`. See for example:

https://github.com/BCCDC-PHL/tb-db/blob/e0e8beb99417982503741ad9a25c10d2987991c9/tb_db/crud.py#L245-L250

Could you add a unit test to check how the `sample.id` behaves when a sample is added to a cgMLST cluster? I'm not sure if the existing record will be invalidated, so the valid `sample.id` may be changed when it's added to the cgMLST cluster. Does the returned value have the latest valid `sample.id`?

Reply:
I added two lines in test_crud.py line #82 --> self.assertEqual(created_sample.id,1) and line #90-->  self.assertEqual(result.id, 2), which passed the test. it showed that when we first create the sample, it is assigned an id 1, after we add this sample to a cgmlst cluster, the record with id 1 got invalidated and now it is given a new id, 2.